components/zip321/src/lib.rs

```diff
@@ -8,108 +8,68 @@ use core::fmt::Debug;
 use std::{
     collections::BTreeMap,
     fmt::{self, Display},
-};
-
-use base64::{prelude::BASE64_URL_SAFE_NO_PAD, Engine};
-use nom::{
-    character::complete::char, combinator::all_consuming, multi::separated_list0,
-    sequence::preceded,
+    ops::{Deref, DerefMut},
 };
 
 use zcash_address::{ConversionError, ZcashAddress};
 use zcash_protocol::{
     memo::{self, MemoBytes},
-    value::BalanceError,
-    value::Zatoshis,
+    value::{BalanceError, Zatoshis, COIN},
 };
 
 /// Errors that may be produced in decoding of payment requests.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Zip321Error {
-    /// A memo field in the ZIP 321 URI was not properly base-64 encoded
-    InvalidBase64(base64::DecodeError),
     /// A memo value exceeded 512 bytes in length or could not be interpreted as a UTF-8 string
     /// when using a valid UTF-8 lead byte.
     MemoBytesError(memo::Error),
-    /// The ZIP 321 request included more payments than can be created within a single Zcash
-    /// transaction. The wrapped value is the number of payments in the request.
-    TooManyPayments(usize),
-    /// Parsing encountered a duplicate ZIP 321 URI parameter for the returned payment index.
-    DuplicateParameter(parse::Param, usize),
     /// The payment at the wrapped index attempted to include a memo when sending to a
     /// transparent recipient address, which is not supported by the protocol.
     TransparentMemo(usize),
-    /// The payment at the wrapped index did not include a recipient address.
-    RecipientMissing(usize),
     /// The ZIP 321 URI was malformed and failed to parse.
-    ParseError(String),
+    ParseError(zip321_parse::Error),
+    /// Any other stringly typed error
+    Other(String),
```

Comment:
I split `ParseError` up into two variants - most have moved to `zip321_parser` but a few validation messages remain.